### ISSUE TITLE
Maybe a big bug in object_detection API, when I add data augmentation options on config files, then the model quickly divergences, mAP  on testset would be 0. 

### ISSUE THREAD

**Issue by MetaPeak** [tensorflow/models#2066]
I am training with "ssd_inception_v2" on coco datasets, when I use the default data augmentation options in config file, which is 
```
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  data_augmentation_options {
    ssd_random_crop {
    }
  }
```
the model convergences rightly. It works very well. However, when I add one data augmentation options which is defined in `objection_detection/coco/preprocessor`, only add one `random_adjust_hue` option or  
`random_adjust_saturation` et.al. The model would  divergence quickly, even in a pretrained model， and mAP on testset would decline to 0 quickly. I add data augmentation options on config files, which looks like this format
```
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  data_augmentation_options {
    ssd_random_crop {
    }
  }

  data_augmentation_options {
    random_adjust_saturation {
    }
  }
```
![image](https://user-images.githubusercontent.com/25903017/28741973-5c8b0358-7455-11e7-9e1e-7a2ad0a221e2.png)
**I tried `ssd_inception_v2`  on pascal data with the all scripts in Tensorflow Object Detection API, I met the same  problem. Without `random_adjust_brightness` or any of hue, contrast. et.al, the model works quite well, when add `random_adjust_brightness` or any of hue, contrast. et.al. It's mAP on validation set quickly decline to 0. I wonder whether those methods on `core/preporcessor.py` loss the bboxes or labels when they are applied?**

**Comment by MetaPeak** [comment]
@jch1  @ali01

**Comment by shivaniag** [comment]
@derekjchow could you please take a look.

**Comment by drauh** [comment]
This is because of the clipping [there](https://github.com/tensorflow/models/blob/3d792f935d652b2c7793b95aa3351a5551dc2401/object_detection/core/preprocessor.py#L510). It seems that models need input to be in fixed point representation (see the [preprocess function](https://github.com/tensorflow/models/blob/3d792f935d652b2c7793b95aa3351a5551dc2401/object_detection/models/ssd_inception_v2_feature_extractor.py#L45)).

In fact, all tf.image.random_* function make sure the output is valid so this clip operation is not needed. I removed them and made a pull request here : #2146.

**Comment by MetaPeak** [comment]
@drauh I remove the function `tf.clip_by_value`， the model indeed converges. However, if we remove `tf_clip_by_value`, the value of the image would be more than 1 or less than 0,  how to fix this problem. It is amazing, why we add the `tf.clip_by_value` funciton, the model wouldn't converge?  It is unreasonable~
```
In [153]: img = np.random.random((2,2,3))

In [154]: img
Out[154]: 
array([[[ 0.9787129 ,  0.66530407,  0.07759292],
        [ 0.07755117,  0.8064815 ,  0.10519789]],

       [[ 0.15906181,  0.28527012,  0.40126713],
        [ 0.75341048,  0.0888517 ,  0.42916807]]])

In [155]: img1 = tf.image.random_brightness(img, max_delta=0.8)

In [156]: img1 = sess.run(img1)

In [157]: img1
Out[157]: 
array([[[ 0.37353563,  0.06012678, -0.52758437],
        [-0.5276261 ,  0.2013042 , -0.49997938]],

       [[-0.44611549, -0.31990716, -0.20391014],
        [ 0.14823318, -0.51632559, -0.17600921]]])

In [158]: 


```

**Comment by drauh** [comment]
If the input is in true fixed point representation (if you look at the create_*_tf_record.py scripts, image are encoded as bytes = uint8 if I am not mistaken. So at least at the beginning the are in fixed point), then the [adjust_*](https://www.tensorflow.org/api_docs/python/tf/image/adjust_brightness) functions take care of the conversion to float, apply adjustment and then convert them back to fixed point (and also saturate them). So the delta is not applied in fixed point representation. Maybe you can try max_delta=5, if it is applied in fixed point it will not change much but if is it in float it will prevent the learning.

**Comment by drauh** [comment]
Also on the saturation : if the input (which never append I object_detection api) is in float then it will never be saturated by the the adjust_* functions. Maybe that's a problem to have different behavior to two dtypes. One solution would be to convert everything to fixed point before doing anything since the preprocessing functions assume that dtype (or more precisely that scale, the type is not important).

**Comment by irmowan** [comment]
I have encountered the same problem. I used your MR, https://github.com/tensorflow/models/pull/2146/files, and the model works.
Thank you so much! @drauh

**Comment by sczhengyabin** [comment]
@irmowan Does the extra data augmentation steps works better?

**Comment by irmowan** [comment]
@sczhengyabin For my experiments, the improvement of performance with distort color steps is minimal, just FYI.

**Comment by TKassis** [comment]
I noticed this problem too, adding some augmentation options causes my training to stop converging.

**Comment by CODEJY** [comment]
if num_train_sets = batch_size*train_steps,does 'num_train_sets' include the augmentation data?

**Comment by plieningerweb** [comment]
This is a so essential bug, many many people get frustrated by it! Please check the pull request and merge it (if nothing else gets broken of course...)

**Comment by MetaPeak** [comment]
@drauh I wonder the data argumentation methods' order used in tensorflow detectiuon api. Maybe it uses methods in preprocessor.py first, then it normolizes the value from 0~255 to -1~1. If this is true, the methods such as random adjust brightness/contrast is useless, because add a small number to 255 is approximitely equal to 255.

**Comment by lcerman** [comment]
I have hit this issue too. I even tried adding just:

```
  data_augmentation_options {
    random_adjust_brightness {
      max_delta: 0.0001
    }
  }
```

even tried `max_delta: 0.`, which should end up with identity (i.e., no augmentation). The faster-rccn-resnet101 model (initialized from the pre-trained model) that otherwise trains well does not start learning at all, and the mAP/APs remain close to zero. This indicates, that adding the brightness augmentation severely breaks something in the object_detection API.

**Comment by zengyu714** [comment]
Hi all, 
I already remove the clip of `[0, 1]` and the input of `preprocessor.py` is in the range of `[0, 255]`.
Then the output, when I add some random augmentation, would be out of [0, 255], e.g., `[-20, 300]`
Should I **clip** this back to be compatible with `(2.0 / 255.0) * resized_inputs - 1.0` ?

**Comment by funkysandman** [comment]
should a mirror image return the same score as the same regular image when random_horizontal_flip is used?  I've discovered that if I flip a test image i get a dramatically different score when detecting

**Comment by bugerry87** [comment]
Is this the issue for a behaviour like this?:
```
INFO:tensorflow:global step 265: loss = 0.1965 (0.155 sec/step)
INFO:tensorflow:global step 266: loss = 0.1991 (0.155 sec/step)
INFO:tensorflow:global step 267: loss = 0.1896 (0.154 sec/step)
INFO:tensorflow:global step 268: loss = 0.1847 (0.154 sec/step)
INFO:tensorflow:global step 269: loss = 0.1858 (0.153 sec/step)
INFO:tensorflow:global step 270: loss = 0.1864 (0.151 sec/step)
INFO:tensorflow:global step 271: loss = 0.1861 (0.158 sec/step)
INFO:tensorflow:global step 272: loss = 0.2479 (0.158 sec/step)
INFO:tensorflow:global step 273: loss = 0.3874 (0.159 sec/step)
INFO:tensorflow:global step 274: loss = 0.2599 (0.158 sec/step)
INFO:tensorflow:global step 275: loss = 27207767073038008320.0000 (0.155 sec/step)
INFO:tensorflow:global step 276: loss = 363770730445224804352.0000 (0.154 sec/step)
INFO:tensorflow:global step 277: loss = 2319587573063963639808.0000 (0.157 sec/step)
INFO:tensorflow:global step 278: loss = 9538479895582634672128.0000 (0.155 sec/step)
INFO:tensorflow:global step 279: loss = 35610680577759077466112.0000 (0.153 sec/step)
INFO:tensorflow:global step 280: loss = 93008970008403315261440.0000 (0.153 sec/step)
```
I train a faster-rcnn +  resnet + coco + (my dataset) in summer 2017 and worked well. Nowadays after some updates I get THIS.

**Comment by jashshopin** [comment]
@pkulzc have these changes been made in your repository which is gcp 1.2 ready? I am getting a similar error.

**Comment by pkulzc** [comment]
Sorry they are not merged to my branch. My branch is for temporary use only and not guaranteed to sync. 
We are migrating to new interfaces with can be run on CMLE runtime 1.6 very soon. So you can edit your own repo as a temp fix and use our new interface later on.

**Comment by aysark** [comment]
Subbing, cause i might be running into this issue.

**Comment by kerolos** [comment]
hello @bugerry87 , I have the same issue, could you please inform me how can I solve this strange behavior of the model. 
Thanks in advance.

**Comment by MetaPeak** [comment]
@kerolos Maybe you are using and old version of object_dection, the new version has fixed the bugs. In old version of `core/preprocessor.py` , it has `image = tf.clip_by_value(image, clip_value_min=0, clip_value_max=1)` , you can change the clip_value_max to 255 to avoid the problem.
